apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,921 @@
+    previous: 'thread-list',
+    alsoContains: ['report-view', 'composer', 'group-view']
+  }),
+  Object.freeze({
+    name: 'thread-list',
+    url: 'views/inbox/index.html',
+    behavior: 'InboxView'
+  }),
+  Object.freeze({
+    name: 'report-view',
+    behavior: 'ReportView',
+    previous: 'thread',
+    partOf: 'thread'
+  }),
+  Object.freeze({
+    name: 'composer',
+    behavior: 'ConversationView',
+    previous: 'thread-list',
+    partOf: 'thread'
+  }),
+  Object.freeze({
+    name: 'group-view',
+    behavior: 'GroupView',
+    previous: 'thread',
+    partOf: 'thread'
+  })
+]);
+
+/**
+  * OA_VIEWS holds the configuration for the old non-split architecture.
+  */
+const OA_VIEWS = Object.freeze([
+  Object.freeze({
+    name: 'thread-list',
+    url: 'index.html',
+    behavior: 'InboxView',
+    alsoContains: ['thread', 'composer', 'group-view', 'report-view']
+  }),
+  Object.freeze({
+    name: 'thread',
+    behavior: 'ConversationView',
+    previous: 'thread-list',
+    partOf: 'thread-list'
+  }),
+  Object.freeze({
+    name: 'composer',
+    behavior: 'ConversationView',
+    previous: 'thread-list',
+    partOf: 'thread-list'
+  }),
+  Object.freeze({
+    name: 'group-view',
+    behavior: 'GroupView',
+    previous: 'thread',
+    partOf: 'thread-list'
+  }),
+  Object.freeze({
+    name: 'report-view',
+    behavior: 'ReportView',
+    previous: 'thread',
+    partOf: 'thread-list'
+  })
+]);
+
+const VIEWS = isUsingOldArchitecture() ? OA_VIEWS : NGA_VIEWS;
+
+var currentState;
+var navigationTransition;
+
+/**
+ * Navigate to this URL.
+ *
+ * @param {String} url URL to load.
+ */
+function setLocation(url) {
+  debug('Setting location', url);
+  window.location.assign(url);
+}
 
-// This function is used as a rejection handler, to swallow errors and log them
-function catchError(e) {
-  if (e) {
+/**
+ * We're using the old architecture if we're not in a split document.
+ *
+ * @returns {Boolean} true if we're using the old architecture.
+ */
+function isUsingOldArchitecture() {
+  debug('isUsingOldArchitecture()');
+  var pathname = window.location.pathname;
+  var result = (pathname === '/' || pathname === '/index.html');
+  debug('isUsingOldArchitecture() =>', result);
+  return result;
+}
+
+/**
+ * Parses hash to find a view name.
+ * It takes the part of the hash between '/' or the start of the string, and '?'
+ * or the end of the string.
+ *
+ * Examples:
+ *   #/conversation?id=5               => conversation
+ *   #/composer?recipient=+33615647897 => composer
+ *   #composer                          => composer
+ *
+ * @param {String} hash The string to analyze.
+ * @returns {String} The analyzed string. It is not guaranteed to be an existing
+ * view name.
+ */
+function findViewNameFromHash(hash) {
+  debug('findViewNameFromHash(%s)', hash);
+  if (!hash) {
+    return null;
+  }
+
+  if (hash[0] === '#') {
+    hash = hash.slice(1);
+  }
+
+  var index = hash.indexOf('/');
```

Comment:
My main intention was to reduce to __1__ both number of possible supported formats (only string that always starts from "#/" - actually I like it :)) and number of slice calls.

Agree that magic numbers aren't good, but looks like we'll still have at least one either "1" or "2" when we call slice and I personally would try to avoid regex.

So to summarize, the most important thing for me here is to have as less supported string format as possible, the rest is just bikeshedding :)